src/ContaoCommunityAlliance/Composer/Plugin/Environment/ContaoEnvironmentFactory.php

```diff
@@ -0,0 +1,80 @@
+<?php
+
+namespace ContaoCommunityAlliance\Composer\Plugin\Environment;
+
+use Composer\Composer;
+use Composer\Package\RootPackageInterface;
+
+class ContaoEnvironmentFactory
+{
+
+    /**
+     * @param Composer $composer
+     *
+     * @return ContaoEnvironmentInterface
+     */
+    public function create(Composer $composer)
+    {
+        $rootDir = $this->findRoot($composer->getPackage());
+
+        if ($this->isContao3($rootDir)) {
+            return new Contao3Environment($rootDir);
+        }
+
+        if ($this->isContao4($rootDir)) {
+            return new Contao4Environment($rootDir, $composer);
+        }
+
+        throw new UnknownEnvironmentException('Contao installation was not found.');
+    }
+
+    private function isContao3($rootDir)
+    {
+        return is_file($rootDir . '/system/config/constants.php');
+    }
+
+    private function isContao4($rootDir)
+    {
+        return is_file($rootDir . '/app/console');
```

Comment:
You should probably check for `$rootDir . '/vendor/contao/core-bundle` instead, because you don't know if the final app contains the `app/console` script.